bbq/compiler/compiler.go

```diff
@@ -591,17 +622,86 @@ func (c *Compiler[_, _]) compileDeclaration(declaration ast.Declaration) {
 	ast.AcceptDeclaration[struct{}](declaration, c)
 }
 
-func (c *Compiler[_, _]) compileBlock(block *ast.Block) {
+func (c *Compiler[_, _]) compileBlock(block *ast.Block, enclosingDeclKind common.DeclarationKind) {
 	locals := c.currentFunction.locals
 	locals.PushNewWithCurrent()
 	defer locals.Pop()
 
-	for _, statement := range block.Statements {
-		c.compileStatement(statement)
+	if c.shouldPatchReturns(enclosingDeclKind) {
+		c.pushReturns()
```

Comment:
The push is gated by `shouldPatchReturns`, but the pop below is gated by effectively `(enclosingDeclKind == common.DeclarationKindFunction || enclosingDeclKind == common.DeclarationKindInitializer) && hasPostConditions`. That matches `shouldPatchReturns` at the moment, but might get out-of-sync. Maybe ensure the corresponding pop for the push with a `defer`

Reply:
Good idea! Added in https://github.com/onflow/cadence/pull/3816/commits/fc87b300480b695dd05bfd2ea0f267fa47ff8a84